[IAP] Make `UpgradesViewModel` testable

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: #9884 . Branched of https://github.com/woocommerce/woocommerce-ios/pull/9907, which should be merged first
<!-- Id number of the GitHub issue this PR addresses. -->

## Description
This PR makes `UpgradesViewModel` more testable by declaring a `InAppPurchasesForWPComPlansProtocol` dependency in its initializer, which defaults to an instance of `InAppPurchasesForWPComPlansManager` if none is given

## Changes
- We remove the class-level `@MainActor` declaration in `InAppPurchasesForWPComPlansManager` in favor of declaring it per-method basis. This is done so we can inject objects that conform to the protocol without the constraint of decorating them as `@MainActor` as well (for example when mocking), since we only need the current methods to run in the main thread due being async wrappers for existing `InAppPurchaseAction`s
- Add tests for fetchPlans(), and retrievePlanDetailsIfAvailable() methods. More tests will be added when we handle the result of purchasePlan() and the different errors.

## Testing instructions
- Upgrading a Free trial store should work
- Tests should pass

iamgabrielma committed:
Inject IAPforWPComPlansManager into UpgradesViewModel

iamgabrielma committed:
Rename MockInAppPurchases. Remove Debug flags

iamgabrielma committed:
Add multiple debug plans to MockIAPManager

iamgabrielma committed:
Update access control. Rename load to fetch

iamgabrielma committed:
Update unit tests

joshheald commented on WooCommerce/Classes/Authentication/Store Creation/Plan/MockInAppPurchasesForWPComPlansManager.swift:
```diff
@@ -1,10 +1,8 @@
 
-#if DEBUG
-
 /// Only used during store creation development before IAP server side is ready.
```
Given that this is a mock, but in the main target, it seems reasonable to keep it behind an `#if DEBUG` check. Was there a specific reason you needed to remove it?

Alternatively, it could be moved to `WooCommerceTests/Mocks`

joshheald replied:
I've moved it in 158bebf

iamgabrielma replied:
Thanks! Seems better placed in the test target indeed

joshheald committed:
Fix warning

joshheald committed:
Fix typo

joshheald committed:
9884 Fix UpgradesViewModel tests to use mock

The mock IAPManager wasn’t used by all the tests, and when we call `sut.fetchPlans` without injecting the mock, `inAppPurchasesAreSupported` gets called and leaks its continuation. This may be related to using a testing Stores manager and dispatching actions to it, but I didn’t dig in to that.

This change avoids us testing any of these methods on the real IAP manager.

Additionally, we’ve made the injection a bit clearer by using obvious test data.

joshheald committed:
9884 Move mock to tests target

joshheald commented on WooCommerce/Classes/ViewModels/InAppPurchases/UpgradesViewModel.swift:
```diff
@@ -84,6 +64,28 @@ final class UpgradesViewModel: ObservableObject {
+    }
+}
+
 extension UpgradesViewModel {
     enum AvailableInAppPurchasesWPComPlans: String {
         case essentialMonthly = "debug.woocommerce.express.essential.monthly"
```
Not this PR,  but I noticed it. I don't _think_ we'll want this in the long term, rather to fetch them from the IAP endpoint only. I've worked round it in the tests for now.

iamgabrielma replied:
Logged here: https://github.com/woocommerce/woocommerce-ios/issues/9927

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Upgrades/UpgradesViewModelTests.swift:
```diff
@@ -5,34 +5,77 @@ final class UpgradesViewModelTests: XCTestCase {
+        let sut = UpgradesViewModel(siteID: sampleSiteID)
+
+        // When
+        await sut.fetchPlans()
```
Not using the mock here causes the tests to hang, because `inAppPurchasesAreSupported` on the real IAP manager leaks its continuation when used in the tests. I've swapped it to using the mock in c248780, correctly configured.

joshheald commented on WooCommerce/WooCommerceTests/ViewRelated/Upgrades/UpgradesViewModelTests.swift:
```diff
@@ -5,34 +5,77 @@ final class UpgradesViewModelTests: XCTestCase {
+
+        // When
+        await sut.fetchPlans()
+        XCTAssertEqual(sut.wpcomPlans.first?.displayName, "Debug Monthly", "Precondition")
```
Nice – I didn't know about the note parameter

joshheald reviewed (approved):
I've made a few changes to the tests, but good job making it more testable here 😊 

Notes in the comments and commit messages, but give me a shout if you want to discuss anything.
